controllers/vsphereclusteridentity_controller.go

```diff
@@ -181,9 +189,16 @@ func (r clusterIdentityReconciler) reconcileDelete(ctx _context.Context, identit
 	if ctrlutil.ContainsFinalizer(secret, legacyIdentityFinalizer) {
 		ctrlutil.RemoveFinalizer(secret, legacyIdentityFinalizer)
 	}
+
+	// TODO (killianmuldoon): Consider whether we should have a finalizer on the secret and, if so, whether this finalizer should be renamed.
 	ctrlutil.RemoveFinalizer(secret, infrav1.SecretIdentitySetFinalizer)
 	if err := r.Client.Update(ctx, secret); err != nil {
 		return err
 	}
-	return r.Client.Delete(ctx, secret)
+	if err := r.Client.Delete(ctx, secret); err != nil {
```

Comment:
It's not clear to me that there has to be a 1:1:1 relationship between the secret and the VSphereClusterIdentity and the Cluster, but it looks like the intended behaviour of this controller is to clean up the secret it references.

The secret can't be deleted until the finalizer is removed, but it's the same finalizer that is added by each VSphereClusterIdentity.

I think there's a bunch of weirdness here that IMO is a bug, but may be being relied on as a feature by someone.

Maybe the best course is to open an issue to establish rules for the lifecycle of this object?

Reply:
I think it's not a 1:1:1 relationship based on 
> * Credentials via a VSphereClusterIdentity: `VSphereClusterIdentity` is a cluster-scoped resource and enables multiple VSphereClusters to share the same set of credentials. The namespaces that are allowed to use the VSphereClusterIdentity can also be configured via a `LabelSelector`.
> https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/0300fab21584ad0f026e15b6fb5208708b37ff88/docs/identity_management.md#L12-L11

Reply:
You're right - it's many Clusters to one VSphereClusterIdentity to one Secret. The VSphereClusterIdentity is linked to the CAPV deployment rather than any individual cluster or group of clusters.

The current setup of these finalizers is still odd to me. I think this change is necessary to ensure the vsphereCluster deletes the secret as a cleanup - but do we need the secret to be protected from deletion by a finalizer as well?

Reply:
Based on the proposal (20210311-single-controller-multitenancy.md) I think VSphereClusterIdentity was introduced partially to be re-used for multiple clusters

Reply:
Based on this, is the question then should be allowed to use a single secret in multiple VSphereClusterIdentities?

Reply:
l.141 seems to block that a secret is used by multiple VSphereClusterIdentities

Reply:
> IMO we should deprecate the finalizer on the Secret used by the VSphereClusterIdentity.

Why?

Reply:
My question is why is it there in the first place. 

I don't think we generally have finalizers on our other secrets. And I think putting the finalizer on the bottom node in the dependency graph is an unusual pattern.

Reply:
I would be in favor of getting rid of it, as it seems the whole finalizer flow for this controller never worked as expected. 

The flow before this PR was to (in no necessary order):
- delete the vsphereclusteridentity
- remove the finalizer from the secret manually
- delete the secret


But fine to leave it as it doesn't do any damage anymore AFAICT.

Reply:
I think I"m missing something crucial. I would have thought the following removes the finalizer and deletes the secret:
```go
	ctrlutil.RemoveFinalizer(secret, infrav1.SecretIdentitySetFinalizer)
	if err := r.Client.Update(ctx, secret); err != nil {
		return err
	}
	return r.Client.Delete(ctx, secret)
```

Just because we didn't have a finalizer on the VSphereClusterIdentity it doesn't mean that reconcileDelete wasn't run. It basically just means that if we miss the delete event we don't hit reconcileDelete (e.g. if the controller was down). I would expect that today in 99% of the cases we go through a reconcileDelete. Probably missing something, but I think only in edge cases the secret is orphaned after the VSphereClusterIdentity is gone

Reply:
> I would expect that today in 99% of the cases we go through a reconcileDelete. Probably missing something, but I think only in edge cases the secret is orphaned after the VSphereClusterIdentity is gone

Not sure how it compares, but in my testing this never happened. The secret was always orphaned and the finalizer was always left in place. I'm not sure why as I would have expected the same given the code. My guess is we hit the "not found" case for the `VSphereClusterIdentity` very quickly as there's nothing slowing down its deletion.

Reply:
Ah interesting. That definitely gives us other options. I assumed we have a comparable case to: https://github.com/kubernetes-sigs/cluster-api/issues/6863 (+/- that in the CAPI issue it's fine to miss a delete event if the controller is done)

Reply:
Considering this I think I have to change my opinion in the other direction. I would expect that we 100% hit the notfound case if we don't have a finalizer on the VSphereClusterIdentity (assuming my theory about how informers work is correct)

Reply:
Okay so it should be safe to assume that reconcileDelete was basically never run. Which also means it doesn't change anything if we drop the code there.

What would you like to do?

1. add the VSphereClusterIdentity finalizer in this PR
2. open an issue/PR to get rid of the secret finalizer? (still not sure if we should fix the finalizer handling instead because I don't know for which reasons it was implemented and if we should rather make sure it works instead of dropping it)

Reply:
It does look like we add the finalizer in the case we use the raw secret in the VSphereCluster, so it's not inconsistent within CAPV. I'm fine with keeping it as-is and just merging this PR (once I fix it up).